docs/rfcs/autoops-flag-trigger.md

```diff
@@ -0,0 +1,228 @@
+
+```Grpc
+rpc ListFlagTriggers(ListFlagTriggersRequest) returns (ListFlagTriggersResponse) {}
+```
+
+```proto
+message ListFlagTriggersRequest {
+    string feature_id = 1;
+    string environment_namespace = 2;
+    int32 page = 3;
+    int32 page_size = 4;
+}
+```
+
+#### AutoOps Trigger: WebHook
+
+First, we will provide an endpoint to receive the trigger request, and then we will trigger the feature flag based on
+the trigger request. The trigger request will be sent by the user's system or tools.
+
+The trigger url secret is generated by the AutoOps server, and it will be encrypted from **trigger id**, **flag id**, *
+*action** of the flag trigger and **a random uuid** generated by `"github.com/bucketeer-io/bucketeer/pkg/uuid"`.
+
+* How to generate the trigger url secret?
+
+Use the following struct to as secret data:
+
+```go
+package flagtrigger
+
+type FlagTriggerSecret struct {
+	ID        string `json:"id"`
+	FeatureID string `json:"feature_id"`
+	Action    int    `json:"action"`
+	UUID      string `json:"uuid"`
+}
+
+```
+
+Then encrypt the secret data by the following code:
+
+```
+trigger_url_secret = crypto.EncrypterDecrypter.Encrypt(FlagTriggerSecret)
+```
+
+* What does the trigger url look like?
+
+```curl
+curl -X POST https://bucketeer.io/trigger/${trigger_url_secret}
+```
+
+* When trigger url been called, how to check the trigger url secret?
+
+First, decrypt the trigger url secret by the following code:
+
+```
+flagTriggerSecret = crypto.EncrypterDecrypter.Decrypt(trigger_url_secret)
+```
+
+Then check the decrypted trigger url secret is valid or not by the following steps:
+
+1. Query database by `flagTriggerSecret.ID` to get the flag trigger data `flagTrigger`.
+2. Check `flagTriggerSecret.FeatureID` is equal to `flagTrigger.FeatureID`, `flagTriggerSecret.Action` is
+   equal to `flagTrigger.Action` and `flagTriggerSecret.UUID` is equal to `flagTrigger.UUID`.
+3. If the above conditions are all true, then call the `feature` API to turn on/off the feature flag base on
+   the `flagTriggerSecret.Action`.
+
+* How to reset the trigger url?
+
+When reset the trigger url, we need to generate a new `uuid`, then update the `flag_trigger`.`uuid` field by the
+new `uuid`, and then encrypt the new `flagTriggerSecret` by the following code:
+
+```
+trigger_url_secret = crypto.EncrypterDecrypter.Encrypt(FlagTriggerSecret)
+```
+
+Then return the new trigger url to the user.
+
+* Only show the trigger url to the user for the first time.
+
+    - When the user creates a new flag trigger, we will generate a new `uuid` and encrypt the `flagTriggerSecret` as the
+      above steps, and then return the trigger url to the user.
+    - When the user reset the trigger url, we will generate a new `uuid` and encrypt the `flagTriggerSecret` as the
+      above steps, and then return the trigger url to the user.
+    - In the `ListFlagTrigger` and `GetFlagTrigger` API response, we will return partial masked trigger url to the user.
```

Comment:
How will you check if it should return a full or partial URL?

Reply:
I think we can return partial URL directly in the `ListFlagTrigger` and `GetFlagTrigger` API response, because we only return the full URL in the `CreateFlagTrigger` API response.